File: Searching and Sorting/merge_sort.py
class Solution:
    def _merge(self, arr, low, high):
        mid = int(low + (high - low)/2)
        left, right = arr[low:mid+1], arr[mid+1:high+1]
        i, j = 0, 0
        merged = []

        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                merged.append(left[i])
                i += 1
            else:
                merged.append(right[j])
                j += 1

        while i < len(left):
            merged.append(left[i])
            i += 1

        while j < len(right):
            merged.append(right[j])
            j += 1

        return merged

    def _merge_sort(self, arr, low, high):
        if low >= high:
            return
        mid = int(low + (high - low)/2)
        self._merge_sort(arr, low, mid)
        self._merge_sort(arr, mid + 1, high)
        arr[low:high+1] = self._merge(arr, low, high)

    def merge_sort(self, arr):
        self._merge_sort(arr, 0, len(arr) - 1)


arr1 = [2, 3, 4, 1]
Solution().merge_sort(arr1)
print(arr1)

arr2 = [56, 2, 45]
Solution().merge_sort(arr2)
print(arr2)

arr3 = [5, 6, 3, 2, 1, 6, 7]
Solution().merge_sort(arr3)
print(arr3)

File: Java/Arrays/Problem16/MergeSort.java
package Arrays.Problem16;

import java.util.ArrayList;
import java.util.List;


class InversionCounter {
    public Integer count;

    public InversionCounter(Integer count) {
        this.count = count;
    }
}


public class MergeSort {
    public static <T extends Comparable<T>> Integer getInversionCount(List<T> arr) {
        /*
            Time complexity is O(n * log(n)) and space complexity is O(n).
         */
        int n = arr.size();
        InversionCounter inversionCounter = new InversionCounter(0);
        sort(arr, 0, n - 1, inversionCounter);
        return inversionCounter.count;
    }

    private static <T extends Comparable<T>> void sort(List<T> arr, int low, int high, InversionCounter inversionCounter) {
        if (low >= high) return;
        int mid = (low + (high - low)/2);
        sort(arr, low, mid, inversionCounter);
        sort(arr, mid + 1, high, inversionCounter);
        List<T> merged = merge(arr, low, high, inversionCounter);
        int counter = 0;
        for (int i = low; i <= high; i += 1) {
            arr.set(i, merged.get(counter));
            counter += 1;
        }
    }

    private static <T extends Comparable<T>> List<T> merge(List<T> arr, int low, int high, InversionCounter inversionCounter) {
        int mid = (low + (high - low)/2);
        List<T> left = arr.subList(low, mid + 1), right = arr.subList(mid + 1, high + 1);
        int i = 0, j = 0;
        List<T> merged = new ArrayList<>();

        while (i < left.size() && j < right.size()) {
            if (left.get(i).compareTo(right.get(j)) <= 0) {
                merged.add(left.get(i));
                i += 1;
            } else {
                merged.add(right.get(j));
                inversionCounter.count += (left.size() - i);
                j += 1;
            }
        }

        while (i < left.size()) {
            merged.add(left.get(i));
            i += 1;
        }

        while (j < right.size()) {
            merged.add(right.get(j));
            j += 1;
        }

        return merged;
    }
}
